src/text/line.rs

```diff
@@ -585,36 +555,96 @@ impl Widget for Line<'_> {
+            // the area width, only truncate the left.
+            let skip_width = match self.alignment {
+                Some(Alignment::Center) => (line_width.saturating_sub(area_width)) / 2,
+                Some(Alignment::Right) => line_width.saturating_sub(area_width),
                 Some(Alignment::Left) | None => 0,
             };
-            x = x.saturating_add(offset);
-            self.to_owned()
+            render_spans(&self.spans, area, buf, skip_width);
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
-                break;
-            }
+    }
+}
+
+/// Renders all the spans of the line that should be visible.
+fn render_spans(spans: &[Span], mut area: Rect, buf: &mut Buffer, span_skip_width: usize) {
+    for (span, span_width, offset) in visible_spans(spans, span_skip_width) {
+        area = area.indent_x(offset);
+        if area.is_empty() {
+            break;
         }
+        span.render_ref(area, buf);
+        let span_width = u16::try_from(span_width).unwrap_or(u16::MAX);
+        area = area.indent_x(span_width);
     }
 }
 
+/// Returns an iterator over the spans that are visible
+///
+/// Visibility is determined by the spans that are partially or completely visible taking into
+/// account just the amount of width to skip from the start of the `Line`.
+fn visible_spans<'a>(
+    spans: &'a [Span],
+    mut span_skip_width: usize,
+) -> impl Iterator<Item = (Span<'a>, usize, u16)> {
+    spans
+        .iter()
+        .map(|span| (span, span.width()))
+        // Filter non visible spans out.
+        .filter_map(move |(span, span_width)| {
+            // Ignore spans that are completely before the offset. Decrement `span_skip_width` by
+            // the span width until we find a span that is partially or completely visible.
+            if span_skip_width >= span_width {
+                span_skip_width = span_skip_width.saturating_sub(span_width);
+                return None;
+            }
+
+            // Apply the skip from the start of the span, not the end as the end will be trimmed
+            // when rendering the span to the buffer.
+            let available_width = span_width.saturating_sub(span_skip_width);
+            span_skip_width = 0; // ensure the next span is rendered in full
+            Some((span, span_width, available_width))
+        })
+        .map(|(span, span_width, available_width)| {
+            if span_width <= available_width {
+                // Span is fully visible. Clone here is fast as the underlying content is `Cow`.
+                return (span.clone(), span_width, 0u16);
+            }
```

Comment:
@joshka You were against this as it changes the order execution and the code.

Also return then else case could be just the else and omit the `return`.

Reply:
My heuristic for writing readable code is that a simple guard clause is usually a simplification worth pursuing when it removes a layer of nesting like this (even at the expense of the order that lines will be encountered like this). If this was more than 1 line like this I'd still want it to be handled as an `if { } else { }`

More specifically: layers of nesting and the else statements here cost more to maintain. This is due to the cognitive burden when reading of carrying both branches in your head for the entire block of code.